### ISSUE TITLE
SLH instantiation does not consistently create operator-valued matrices

### ISSUE THREAD

**Issue by danielwe** [mabuchilab/QNET#55]
Say you're instantiating an SLH object, like this:
```python
slh = SLH(
    [[s00, s01],
     [s10, s11]],
    [l0, l1],
    h,
)
```
Here, the ``s00`` etc. may be numbers, sympy expressions, QNET ``ScalarTimesOperator`` instances, and so on. Since we're giving the matrices as lists, the parameters will be processed internally, to the effect of ``slh.S = Matrix([[s00, ...]]) * IdentityOperator`` and so on. Due to the final multiplication, both S and L will be elementwise operator valued, i.e., each element will be a ``ScalarTimesOperator`` instance or similar.

However, let's say I decide to call ``Matrix`` explicitly myself.
```python
slh = SLH(
    Matrix([[s00, s01],
            [s10, s11]]),
    Matrix([l0, l1]),
    h,
)
```
Since the S and L arguments are already ``Matrix`` instances, they will not be processed further, so in effect, I'm setting ``slh.S = Matrix([[s00, ...]])`` and so on. Hence, each of the elements of ``slh.S`` will retain their original types, which may be both scalars (numbers, sympy expressions, etc.) and operators.

Almost everything you can do with the ``slh`` object still works -- any scalar is interpreted as a scalar times the identity operator when required. However, the behavior seems a little inconsistent. Instead of the current code,
```python
        if not isinstance(S, Matrix):
            S = Matrix(S) * IdentityOperator
        ...
```
one could simply do
```python
        if not isinstance(S, Matrix):
            S = Matrix(S)
        S = S * IdentityOperator
        ...
```
or skip the operator multiplication altogether,
```python
        if not isinstance(S, Matrix):
            S = Matrix(S)
        ...
```

The real reason this came to my attention, however, is that the one place things break is when comparing SLH objects, or their components, for equality. As it turns out, the current implementation of ``__eq__`` for certain objects within QNET is not symmetric. More on that in another issue.

**Comment by danielwe** [comment]
The unsymmetric relations alluded to in the final paragraph actually turned out to be a bug in sympy, see https://github.com/sympy/sympy/issues/13078. The point is that, in the example above (using ``slh1`` for the first variant and ``slh2`` for the second), ``slh1 == slh2`` is True, while ``slh2 == slh1`` can be False. 

The explanation is this: when stepping down the hierarchy, we ultimately end up with comparisons of the type ``s00 * IdentityOperator == s00``. This calls ``ScalarTimesOperator.__eq__``, which explicitly supports these cases, and evaluates to True. However, the reflected version winds up at ``s00 == s00 * IdentityOperator``, and if ``s00`` is a sympy object, this dispatches to ``sympy.Basic.__eq__``, which does not support ``ScalarTimesOperator`` objects, and therefore returns False.

The elegant solution would be for ``sympy.Basic.__eq__`` to return ``NotImplemented`` when it cannot make sense of the arguments. Then the comparison would delegate to the reflected method on the other object, ``s00 * IdentityOperator``, and a symmetric relation would be guaranteed.

Until sympy fixes this, however, a simple workaround for us is to always enforce operator valued elements, i.e., use the first suggested alternative above.
```python
        if not isinstance(S, Matrix):
            S = Matrix(S)
        S = S * IdentityOperator
        ...
```

**Comment by goerz** [comment]
Sounds like you got this nailed down pretty much. Feel free to fix this as you see fit and add a test

**Comment by ngedwin98** [comment]
The inconsistent typing for `Matrix` elements caused me some pain as well when PyCalling QNET from Julia! I would welcome the enforcement of operator-valued elements in this case.

**Comment by goerz** [comment]
Wait... I'm actually not on board with requiring that all the elements of a Matrix are Operators. I'm also using matrices in contexts outside of SLH with scalar entries (to do analytic basis transformations, for example). Sorry if I missed the original point of the report - I thought this was only in an SLH context. Maybe we could add an optional parameter to the Matrix constructor that forces conversion of all elements to Operators?

**Edit**: although maybe this is going to be just fine, with Operators. Maybe just do it, and I'll see if it trips me up

**Comment by danielwe** [comment]
Sorry, didn't see the latest comments before pushing my fix. I was also thinking only in the context of SLH objects, so this is what I've implemented. The Matrix class is unchanged.

Generally, I think it makes the most sense to leave Matrix alone. It is a matrix class for elements of any type and purpose, and should ideally not have to know about the operator algebra at all (although I know it must in order to support the ``expand`` and ``all_symbols`` methods).

**Comment by goerz** [comment]
Then we're on the same page!

**Comment by danielwe** [comment]
Just a note that the underlying sympy issues have now been fixed (see sympy/sympy#13091), so equality/inequality comparison between any scaled IdentityOperator and a sympy scalar will be symmetric as soon as people start using the next sympy release.
